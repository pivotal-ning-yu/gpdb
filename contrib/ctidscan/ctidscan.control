# ctidscan extension
comment = 'ctidscan'
default_version = '1.0'
module_pathname = '$libdir/ctidscan'
relocatable = true
