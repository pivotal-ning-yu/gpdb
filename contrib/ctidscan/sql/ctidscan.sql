--
-- Regression Tests for Custom Plan APIs
--

-- construction of test data
SET client_min_messages TO 'warning';

CREATE SCHEMA regtest_custom_scan;

SET search_path TO regtest_custom_scan, public;

CREATE TABLE t1 (
    a   int primary key,
    b   text
);
INSERT INTO t1 (SELECT s, md5(s::text) FROM generate_series(1,400) s);
VACUUM ANALYZE t1;

CREATE TABLE t2 (
    x   int primary key,
    y   text
);
INSERT INTO t2 (SELECT s, md5(s::text)||md5(s::text) FROM generate_series(1,400) s);
VACUUM ANALYZE t2;

RESET client_min_messages;
--
-- Check Plans if no special extension is loaded.
--
EXPLAIN (costs off) SELECT * FROM t1 WHERE a = 40;
EXPLAIN (costs off) SELECT * FROM t1 WHERE b like '%789%';
EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid = '(2,10)'::tid;
EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;

--
-- Plan for same query but ctidscan was loaded
--
LOAD '$libdir/ctidscan';
EXPLAIN (costs off) SELECT * FROM t1 WHERE a = 40;
EXPLAIN (costs off) SELECT * FROM t1 WHERE b like '%789%';
EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid = '(2,10)'::tid;
EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;
EXPLAIN (costs off) SELECT * FROM t1 JOIN t2 ON t1.ctid = t2.ctid WHERE t1.ctid < '(2,10)'::tid AND t2.ctid > '(1,75)'::tid;

SELECT ctid,* FROM t1 WHERE ctid < '(1,20)'::tid;
SELECT ctid,* FROM t1 WHERE ctid > '(4,0)'::tid;
SELECT ctid,* FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;
SELECT t1.ctid,* FROM t1 JOIN t2 ON t1.ctid = t2.ctid WHERE t1.ctid < '(2,10)'::tid AND t2.ctid > '(1,75)'::tid;

PREPARE p1(tid, tid) AS SELECT ctid,* FROM t1
                        WHERE b like '%abc%' AND ctid BETWEEN $1 AND $2;
EXPLAIN (costs off) EXECUTE p1('(5,0)'::tid, '(10,0)'::tid);
EXPLAIN (costs off) EXECUTE p1('(10,0)'::tid, '(5,0)'::tid);

-- Also, EXPLAIN with none-text format
EXPLAIN (costs off, format xml) EXECUTE p1('(0,0)'::tid, '(5,0)'::tid);

-- Test cleanup
DROP SCHEMA regtest_custom_scan CASCADE;
