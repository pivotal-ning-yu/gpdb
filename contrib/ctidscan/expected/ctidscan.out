--
-- Regression Tests for Custom Plan APIs
--
-- construction of test data
SET client_min_messages TO 'warning';
CREATE SCHEMA regtest_custom_scan;
SET search_path TO regtest_custom_scan, public;
CREATE TABLE t1 (
    a   int primary key,
    b   text
);
INSERT INTO t1 (SELECT s, md5(s::text) FROM generate_series(1,400) s);
VACUUM ANALYZE t1;
CREATE TABLE t2 (
    x   int primary key,
    y   text
);
INSERT INTO t2 (SELECT s, md5(s::text)||md5(s::text) FROM generate_series(1,400) s);
VACUUM ANALYZE t2;
RESET client_min_messages;
--
-- Check Plans if no special extension is loaded.
--
EXPLAIN (costs off) SELECT * FROM t1 WHERE a = 40;
           QUERY PLAN           
--------------------------------
 Index Scan using t1_pkey on t1
   Index Cond: (a = 40)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE b like '%789%';
           QUERY PLAN           
--------------------------------
 Seq Scan on t1
   Filter: (b ~~ '%789%'::text)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid = '(2,10)'::tid;
             QUERY PLAN             
------------------------------------
 Tid Scan on t1
   TID Cond: (ctid = '(2,10)'::tid)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;
                            QUERY PLAN                            
------------------------------------------------------------------
 Seq Scan on t1
   Filter: ((ctid >= '(2,115)'::tid) AND (ctid <= '(3,10)'::tid))
(2 rows)

--
-- Plan for same query but ctidscan was loaded
--
LOAD '$libdir/ctidscan';
EXPLAIN (costs off) SELECT * FROM t1 WHERE a = 40;
           QUERY PLAN           
--------------------------------
 Index Scan using t1_pkey on t1
   Index Cond: (a = 40)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE b like '%789%';
           QUERY PLAN           
--------------------------------
 Seq Scan on t1
   Filter: (b ~~ '%789%'::text)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid = '(2,10)'::tid;
             QUERY PLAN             
------------------------------------
 Tid Scan on t1
   TID Cond: (ctid = '(2,10)'::tid)
(2 rows)

EXPLAIN (costs off) SELECT * FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;
                              QUERY PLAN                              
----------------------------------------------------------------------
 Custom Scan (ctidscan) on t1
   Filter: ((ctid >= '(2,115)'::tid) AND (ctid <= '(3,10)'::tid))
   ctid quals: ((ctid >= '(2,115)'::tid) AND (ctid <= '(3,10)'::tid))
(3 rows)

EXPLAIN (costs off) SELECT * FROM t1 JOIN t2 ON t1.ctid = t2.ctid WHERE t1.ctid < '(2,10)'::tid AND t2.ctid > '(1,75)'::tid;
                    QUERY PLAN                    
--------------------------------------------------
 Merge Join
   Merge Cond: (t1.ctid = t2.ctid)
   ->  Sort
         Sort Key: t1.ctid
         ->  Custom Scan (ctidscan) on t1
               Filter: (ctid < '(2,10)'::tid)
               ctid quals: (ctid < '(2,10)'::tid)
   ->  Sort
         Sort Key: t2.ctid
         ->  Custom Scan (ctidscan) on t2
               Filter: (ctid > '(1,75)'::tid)
               ctid quals: (ctid > '(1,75)'::tid)
(12 rows)

SELECT ctid,* FROM t1 WHERE ctid < '(1,20)'::tid;
  ctid   |  a  |                b                 
---------+-----+----------------------------------
 (0,1)   |   1 | c4ca4238a0b923820dcc509a6f75849b
 (0,2)   |   2 | c81e728d9d4c2f636f067f89cc14862c
 (0,3)   |   3 | eccbc87e4b5ce2fe28308fd9f2a7baf3
 (0,4)   |   4 | a87ff679a2f3e71d9181a67b7542122c
 (0,5)   |   5 | e4da3b7fbbce2345d7772b0674a318d5
 (0,6)   |   6 | 1679091c5a880faf6fb5e6087eb1b2dc
 (0,7)   |   7 | 8f14e45fceea167a5a36dedd4bea2543
 (0,8)   |   8 | c9f0f895fb98ab9159f51fd0297e236d
 (0,9)   |   9 | 45c48cce2e2d7fbdea1afc51c7c6ad26
 (0,10)  |  10 | d3d9446802a44259755d38e6d163e820
 (0,11)  |  11 | 6512bd43d9caa6e02c990b0a82652dca
 (0,12)  |  12 | c20ad4d76fe97759aa27a0c99bff6710
 (0,13)  |  13 | c51ce410c124a10e0db5e4b97fc2af39
 (0,14)  |  14 | aab3238922bcc25a6f606eb525ffdc56
 (0,15)  |  15 | 9bf31c7ff062936a96d3c8bd1f8f2ff3
 (0,16)  |  16 | c74d97b01eae257e44aa9d5bade97baf
 (0,17)  |  17 | 70efdf2ec9b086079795c442636b55fb
 (0,18)  |  18 | 6f4922f45568161a8cdf4ad2299f6d23
 (0,19)  |  19 | 1f0e3dad99908345f7439f8ffabdffc4
 (0,20)  |  20 | 98f13708210194c475687be6106a3b84
 (0,21)  |  21 | 3c59dc048e8850243be8079a5c74d079
 (0,22)  |  22 | b6d767d2f8ed5d21a44b0e5886680cb9
 (0,23)  |  23 | 37693cfc748049e45d87b8c7d8b9aacd
 (0,24)  |  24 | 1ff1de774005f8da13f42943881c655f
 (0,25)  |  25 | 8e296a067a37563370ded05f5a3bf3ec
 (0,26)  |  26 | 4e732ced3463d06de0ca9a15b6153677
 (0,27)  |  27 | 02e74f10e0327ad868d138f2b4fdd6f0
 (0,28)  |  28 | 33e75ff09dd601bbe69f351039152189
 (0,29)  |  29 | 6ea9ab1baa0efb9e19094440c317e21b
 (0,30)  |  30 | 34173cb38f07f89ddbebc2ac9128303f
 (0,31)  |  31 | c16a5320fa475530d9583c34fd356ef5
 (0,32)  |  32 | 6364d3f0f495b6ab9dcf8d3b5c6e0b01
 (0,33)  |  33 | 182be0c5cdcd5072bb1864cdee4d3d6e
 (0,34)  |  34 | e369853df766fa44e1ed0ff613f563bd
 (0,35)  |  35 | 1c383cd30b7c298ab50293adfecb7b18
 (0,36)  |  36 | 19ca14e7ea6328a42e0eb13d585e4c22
 (0,37)  |  37 | a5bfc9e07964f8dddeb95fc584cd965d
 (0,38)  |  38 | a5771bce93e200c36f7cd9dfd0e5deaa
 (0,39)  |  39 | d67d8ab4f4c10bf22aa353e27879133c
 (0,40)  |  40 | d645920e395fedad7bbbed0eca3fe2e0
 (0,41)  |  41 | 3416a75f4cea9109507cacd8e2f2aefc
 (0,42)  |  42 | a1d0c6e83f027327d8461063f4ac58a6
 (0,43)  |  43 | 17e62166fc8586dfa4d1bc0e1742c08b
 (0,44)  |  44 | f7177163c833dff4b38fc8d2872f1ec6
 (0,45)  |  45 | 6c8349cc7260ae62e3b1396831a8398f
 (0,46)  |  46 | d9d4f495e875a2e075a1a4a6e1b9770f
 (0,47)  |  47 | 67c6a1e7ce56d3d6fa748ab6d9af3fd7
 (0,48)  |  48 | 642e92efb79421734881b53e1e1b18b6
 (0,49)  |  49 | f457c545a9ded88f18ecee47145a72c0
 (0,50)  |  50 | c0c7c76d30bd3dcaefc96f40275bdc0a
 (0,51)  |  51 | 2838023a778dfaecdc212708f721b788
 (0,52)  |  52 | 9a1158154dfa42caddbd0694a4e9bdc8
 (0,53)  |  53 | d82c8d1619ad8176d665453cfb2e55f0
 (0,54)  |  54 | a684eceee76fc522773286a895bc8436
 (0,55)  |  55 | b53b3a3d6ab90ce0268229151c9bde11
 (0,56)  |  56 | 9f61408e3afb633e50cdf1b20de6f466
 (0,57)  |  57 | 72b32a1f754ba1c09b3695e0cb6cde7f
 (0,58)  |  58 | 66f041e16a60928b05a7e228a89c3799
 (0,59)  |  59 | 093f65e080a295f8076b1c5722a46aa2
 (0,60)  |  60 | 072b030ba126b2f4b2374f342be9ed44
 (0,61)  |  61 | 7f39f8317fbdb1988ef4c628eba02591
 (0,62)  |  62 | 44f683a84163b3523afe57c2e008bc8c
 (0,63)  |  63 | 03afdbd66e7929b125f8597834fa83a4
 (0,64)  |  64 | ea5d2f1c4608232e07d3aa3d998e5135
 (0,65)  |  65 | fc490ca45c00b1249bbe3554a4fdf6fb
 (0,66)  |  66 | 3295c76acbf4caaed33c36b1b5fc2cb1
 (0,67)  |  67 | 735b90b4568125ed6c3f678819b6e058
 (0,68)  |  68 | a3f390d88e4c41f2747bfa2f1b5f87db
 (0,69)  |  69 | 14bfa6bb14875e45bba028a21ed38046
 (0,70)  |  70 | 7cbbc409ec990f19c78c75bd1e06f215
 (0,71)  |  71 | e2c420d928d4bf8ce0ff2ec19b371514
 (0,72)  |  72 | 32bb90e8976aab5298d5da10fe66f21d
 (0,73)  |  73 | d2ddea18f00665ce8623e36bd4e3c7c5
 (0,74)  |  74 | ad61ab143223efbc24c7d2583be69251
 (0,75)  |  75 | d09bf41544a3365a46c9077ebb5e35c3
 (0,76)  |  76 | fbd7939d674997cdb4692d34de8633c4
 (0,77)  |  77 | 28dd2c7955ce926456240b2ff0100bde
 (0,78)  |  78 | 35f4a8d465e6e1edc05f3d8ab658c551
 (0,79)  |  79 | d1fe173d08e959397adf34b1d77e88d7
 (0,80)  |  80 | f033ab37c30201f73f142449d037028d
 (0,81)  |  81 | 43ec517d68b6edd3015b3edc9a11367b
 (0,82)  |  82 | 9778d5d219c5080b9a6a17bef029331c
 (0,83)  |  83 | fe9fc289c3ff0af142b6d3bead98a923
 (0,84)  |  84 | 68d30a9594728bc39aa24be94b319d21
 (0,85)  |  85 | 3ef815416f775098fe977004015c6193
 (0,86)  |  86 | 93db85ed909c13838ff95ccfa94cebd9
 (0,87)  |  87 | c7e1249ffc03eb9ded908c236bd1996d
 (0,88)  |  88 | 2a38a4a9316c49e5a833517c45d31070
 (0,89)  |  89 | 7647966b7343c29048673252e490f736
 (0,90)  |  90 | 8613985ec49eb8f757ae6439e879bb2a
 (0,91)  |  91 | 54229abfcfa5649e7003b83dd4755294
 (0,92)  |  92 | 92cc227532d17e56e07902b254dfad10
 (0,93)  |  93 | 98dce83da57b0395e163467c9dae521b
 (0,94)  |  94 | f4b9ec30ad9f68f89b29639786cb62ef
 (0,95)  |  95 | 812b4ba287f5ee0bc9d43bbf5bbe87fb
 (0,96)  |  96 | 26657d5ff9020d2abefe558796b99584
 (0,97)  |  97 | e2ef524fbf3d9fe611d5a8e90fefdc9c
 (0,98)  |  98 | ed3d2c21991e3bef5e069713af9fa6ca
 (0,99)  |  99 | ac627ab1ccbdb62ec96e702f07f6425b
 (0,100) | 100 | f899139df5e1059396431415e770c6dd
 (0,101) | 101 | 38b3eff8baf56627478ec76a704e9b52
 (0,102) | 102 | ec8956637a99787bd197eacd77acce5e
 (0,103) | 103 | 6974ce5ac660610b44d9b9fed0ff9548
 (0,104) | 104 | c9e1074f5b3f9fc8ea15d152add07294
 (0,105) | 105 | 65b9eea6e1cc6bb9f0cd2a47751a186f
 (0,106) | 106 | f0935e4cd5920aa6c7c996a5ee53a70f
 (0,107) | 107 | a97da629b098b75c294dffdc3e463904
 (0,108) | 108 | a3c65c2974270fd093ee8a9bf8ae7d0b
 (0,109) | 109 | 2723d092b63885e0d7c260cc007e8b9d
 (0,110) | 110 | 5f93f983524def3dca464469d2cf9f3e
 (0,111) | 111 | 698d51a19d8a121ce581499d7b701668
 (0,112) | 112 | 7f6ffaa6bb0b408017b62254211691b5
 (0,113) | 113 | 73278a4a86960eeb576a8fd4c9ec6997
 (0,114) | 114 | 5fd0b37cd7dbbb00f97ba6ce92bf5add
 (0,115) | 115 | 2b44928ae11fb9384c4cf38708677c48
 (0,116) | 116 | c45147dee729311ef5b5c3003946c48f
 (0,117) | 117 | eb160de1de89d9058fcb0b968dbbbd68
 (0,118) | 118 | 5ef059938ba799aaa845e1c2e8a762bd
 (0,119) | 119 | 07e1cd7dca89a1678042477183b7ac3f
 (0,120) | 120 | da4fb5c6e93e74d3df8527599fa62642
 (1,1)   | 121 | 4c56ff4ce4aaf9573aa5dff913df997a
 (1,2)   | 122 | a0a080f42e6f13b3a2df133f073095dd
 (1,3)   | 123 | 202cb962ac59075b964b07152d234b70
 (1,4)   | 124 | c8ffe9a587b126f152ed3d89a146b445
 (1,5)   | 125 | 3def184ad8f4755ff269862ea77393dd
 (1,6)   | 126 | 069059b7ef840f0c74a814ec9237b6ec
 (1,7)   | 127 | ec5decca5ed3d6b8079e2e7e7bacc9f2
 (1,8)   | 128 | 76dc611d6ebaafc66cc0879c71b5db5c
 (1,9)   | 129 | d1f491a404d6854880943e5c3cd9ca25
 (1,10)  | 130 | 9b8619251a19057cff70779273e95aa6
 (1,11)  | 131 | 1afa34a7f984eeabdbb0a7d494132ee5
 (1,12)  | 132 | 65ded5353c5ee48d0b7d48c591b8f430
 (1,13)  | 133 | 9fc3d7152ba9336a670e36d0ed79bc43
 (1,14)  | 134 | 02522a2b2726fb0a03bb19f2d8d9524d
 (1,15)  | 135 | 7f1de29e6da19d22b51c68001e7e0e54
 (1,16)  | 136 | 42a0e188f5033bc65bf8d78622277c4e
 (1,17)  | 137 | 3988c7f88ebcb58c6ce932b957b6f332
 (1,18)  | 138 | 013d407166ec4fa56eb1e1f8cbe183b9
 (1,19)  | 139 | e00da03b685a0dd18fb6a08af0923de0
(139 rows)

SELECT ctid,* FROM t1 WHERE ctid > '(4,0)'::tid;
 ctid | a | b 
------+---+---
(0 rows)

SELECT ctid,* FROM t1 WHERE ctid BETWEEN '(2,115)'::tid AND '(3,10)'::tid;
  ctid   |  a  |                b                 
---------+-----+----------------------------------
 (2,115) | 355 | 82cec96096d4281b7c95cd7e74623496
 (2,116) | 356 | 6c524f9d5d7027454a783c841250ba71
 (2,117) | 357 | fb7b9ffa5462084c5f4e7e85a093e6d7
 (2,118) | 358 | aa942ab2bfa6ebda4840e7360ce6e7ef
 (2,119) | 359 | c058f544c737782deacefa532d9add4c
 (2,120) | 360 | e7b24b112a44fdd9ee93bdf998c6ca0e
 (3,1)   | 361 | 52720e003547c70561bf5e03b95aa99f
 (3,2)   | 362 | c3e878e27f52e2a57ace4d9a76fd9acf
 (3,3)   | 363 | 00411460f7c92d2124a67ea0f4cb5f85
 (3,4)   | 364 | bac9162b47c56fc8a4d2a519803d51b3
 (3,5)   | 365 | 9be40cee5b0eee1462c82c6964087ff9
 (3,6)   | 366 | 5ef698cd9fe650923ea331c15af3b160
 (3,7)   | 367 | 05049e90fa4f5039a8cadc6acbb4b2cc
 (3,8)   | 368 | cf004fdc76fa1a4f25f62e0eb5261ca3
 (3,9)   | 369 | 0c74b7f78409a4022a2c4c5a5ca3ee19
 (3,10)  | 370 | d709f38ef758b5066ef31b18039b8ce5
(16 rows)

SELECT t1.ctid,* FROM t1 JOIN t2 ON t1.ctid = t2.ctid WHERE t1.ctid < '(2,10)'::tid AND t2.ctid > '(1,75)'::tid;
  ctid  |  a  |                b                 |  x  |                                y                                 
--------+-----+----------------------------------+-----+------------------------------------------------------------------
 (1,76) | 196 | 084b6fbb10729ed4da8c3d3f5a3ae7c9 | 157 | 6c4b761a28b734fe93831e3fb400ce876c4b761a28b734fe93831e3fb400ce87
 (1,77) | 197 | 85d8ce590ad8981ca2c8286f79f59954 | 158 | 06409663226af2f3114485aa4e0a23b406409663226af2f3114485aa4e0a23b4
 (1,78) | 198 | 0e65972dce68dad4d52d063967f0a705 | 159 | 140f6969d5213fd0ece03148e62e461e140f6969d5213fd0ece03148e62e461e
 (1,79) | 199 | 84d9ee44e457ddef7f2c4f25dc8fa865 | 160 | b73ce398c39f506af761d2277d853a92b73ce398c39f506af761d2277d853a92
 (1,80) | 200 | 3644a684f98ea8fe223c713b77189a77 | 161 | bd4c9ab730f5513206b999ec0d90d1fbbd4c9ab730f5513206b999ec0d90d1fb
 (1,81) | 201 | 757b505cfd34c64c85ca5b5690ee5293 | 162 | 82aa4b0af34c2313a562076992e50aa382aa4b0af34c2313a562076992e50aa3
 (2,1)  | 241 | f340f1b1f65b6df5b5e3f94d95b11daf | 163 | 0777d5c17d4066b82ab86dff8a46af6f0777d5c17d4066b82ab86dff8a46af6f
 (2,2)  | 242 | e4a6222cdb5b34375400904f03d8e6a5 | 164 | fa7cdfad1a5aaf8370ebeda47a1ff1c3fa7cdfad1a5aaf8370ebeda47a1ff1c3
 (2,3)  | 243 | cb70ab375662576bd1ac5aaf16b3fca4 | 165 | 9766527f2b5d3e95d4a733fcfb77bd7e9766527f2b5d3e95d4a733fcfb77bd7e
 (2,4)  | 244 | 9188905e74c28e489b44e954ec0b9bca | 166 | 7e7757b1e12abcb736ab9a754ffb617a7e7757b1e12abcb736ab9a754ffb617a
 (2,5)  | 245 | 0266e33d3f546cb5436a10798e657d97 | 167 | 5878a7ab84fb43402106c575658472fa5878a7ab84fb43402106c575658472fa
 (2,6)  | 246 | 38db3aed920cf82ab059bfccbd02be6a | 168 | 006f52e9102a8d3be2fe5614f42ba989006f52e9102a8d3be2fe5614f42ba989
 (2,7)  | 247 | 3cec07e9ba5f5bb252d13f5f431e4bbb | 169 | 3636638817772e42b59d74cff571fbb33636638817772e42b59d74cff571fbb3
 (2,8)  | 248 | 621bf66ddb7c962aa0d22ac97d69b793 | 170 | 149e9677a5989fd342ae44213df68868149e9677a5989fd342ae44213df68868
 (2,9)  | 249 | 077e29b11be80ab57e1a2ecabb7da330 | 171 | a4a042cf4fd6bfb47701cbc8a1653adaa4a042cf4fd6bfb47701cbc8a1653ada
(15 rows)

PREPARE p1(tid, tid) AS SELECT ctid,* FROM t1
                        WHERE b like '%abc%' AND ctid BETWEEN $1 AND $2;
EXPLAIN (costs off) EXECUTE p1('(5,0)'::tid, '(10,0)'::tid);
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Custom Scan (ctidscan) on t1
   Filter: ((b ~~ '%abc%'::text) AND (ctid >= '(5,0)'::tid) AND (ctid <= '(10,0)'::tid))
   ctid quals: ((ctid >= '(5,0)'::tid) AND (ctid <= '(10,0)'::tid))
(3 rows)

EXPLAIN (costs off) EXECUTE p1('(10,0)'::tid, '(5,0)'::tid);
                                       QUERY PLAN                                        
-----------------------------------------------------------------------------------------
 Custom Scan (ctidscan) on t1
   Filter: ((b ~~ '%abc%'::text) AND (ctid >= '(10,0)'::tid) AND (ctid <= '(5,0)'::tid))
   ctid quals: ((ctid >= '(10,0)'::tid) AND (ctid <= '(5,0)'::tid))
(3 rows)

-- Also, EXPLAIN with none-text format
EXPLAIN (costs off, format xml) EXECUTE p1('(0,0)'::tid, '(5,0)'::tid);
                                                QUERY PLAN                                                 
-----------------------------------------------------------------------------------------------------------
 <explain xmlns="http://www.postgresql.org/2009/explain">                                                 +
   <Query>                                                                                                +
     <Plan>                                                                                               +
       <Node-Type>Custom Scan</Node-Type>                                                                 +
       <Custom-Plan-Provider>ctidscan</Custom-Plan-Provider>                                              +
       <Relation-Name>t1</Relation-Name>                                                                  +
       <Alias>t1</Alias>                                                                                  +
       <Filter>((b ~~ '%abc%'::text) AND (ctid &gt;= '(0,0)'::tid) AND (ctid &lt;= '(5,0)'::tid))</Filter>+
       <ctid-quals>((ctid &gt;= '(0,0)'::tid) AND (ctid &lt;= '(5,0)'::tid))</ctid-quals>                 +
     </Plan>                                                                                              +
   </Query>                                                                                               +
 </explain>
(1 row)

-- Test cleanup
DROP SCHEMA regtest_custom_scan CASCADE;
NOTICE:  drop cascades to 2 other objects
DETAIL:  drop cascades to table t1
drop cascades to table t2
